solidity/test/ValidatorAnnounce.t.sol

```diff
@@ -0,0 +1,112 @@
+import "../contracts/ValidatorAnnounce.sol";
+import {TypeCasts} from "../contracts/libs/TypeCasts.sol";
+
+contract ValidatorAnnounceTest is Test {
+    using TypeCasts for address;
+
+    // TODO: dedup
+    event ValidatorAnnouncement(
+        address indexed validator,
+        string storageLocation
+    );
+
+    MockMailbox mailbox;
+    uint32 localDomain = 1;
+    ValidatorAnnounce valAnnounce;
+
+    function setUp() public {
+        mailbox = new MockMailbox(localDomain);
+        valAnnounce = new ValidatorAnnounce(address(mailbox));
+    }
+
+    function announce(uint256 privateKey, string memory storageLocation)
+        internal
+    {
+        address validator = vm.addr(privateKey);
+        bytes32 domainHash = keccak256(
+            abi.encodePacked(
+                localDomain,
+                address(mailbox).addressToBytes32(),
+                "HYPERLANE"
+            )
+        );
+        bytes32 digest = ECDSA.toEthSignedMessageHash(
+            keccak256(abi.encodePacked(domainHash, storageLocation))
+        );
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(privateKey, digest);
+        bytes memory signature = abi.encodePacked(r, s, v);
+
+        valAnnounce.announce(validator, storageLocation, signature);
+    }
+
+    function assertEqAddrArr(address[] memory a, address[] memory b) internal {
+        bytes memory encodedA = abi.encodePacked(a);
+        bytes memory encodedB = abi.encodePacked(b);
+        assertEq(encodedA, encodedB);
+    }
+
+    function assertEqStrArrArr(string[][] memory a, string[][] memory b)
+        internal
+    {
+        assertEq(a.length, b.length);
+        bytes memory encodedA = abi.encode(a);
+        bytes memory encodedB = abi.encode(b);
+        assertEq(encodedA, encodedB);
+    }
+
+    function testAnnounce() public {
+        uint256 privateKey = 123456789;
+        // Announce a first location
+        address validator = vm.addr(privateKey);
+        string memory storageLocation1 = "s3://test-bucket/us-east-1";
+        //emit InterchainAccountCreated(originDomain, address(this), ica);
+        vm.expectEmit(true, false, false, true, address(valAnnounce));
+        emit ValidatorAnnouncement(validator, storageLocation1);
+        announce(privateKey, storageLocation1);
+
+        address[] memory expectedValidators = new address[](1);
+        expectedValidators[0] = validator;
+        assertEqAddrArr(
+            valAnnounce.getAnnouncedValidators(),
+            expectedValidators
+        );
+
+        string[][] memory expectedLocations1 = new string[][](1);
+        string[] memory locations1 = new string[](1);
+        locations1[0] = storageLocation1;
+        expectedLocations1[0] = locations1;
+        assertEqStrArrArr(
+            valAnnounce.getAnnouncedStorageLocations(expectedValidators),
+            expectedLocations1
+        );
+
+        // Shouldn't be able to announce the same location twice
```

Comment:
not sure why this is an important invariant but will check the HIP

Reply:
It's not critical, it just makes the relayer's life slightly more convenient. 

Note that since the validator doesn't need to be the one registering the announcement, without replay protection *anyone* could replay announcements to bloat the storageLocations list with duplicate entries.